### ISSUE TITLE
IPv4Range Missing Bits and FiniteBits Instance

### ISSUE THREAD

**Issue by LemonIceScream** [andrewthad/haskell-ip#30]
IPv4Range is a perfect candidate of FiniteBits as it can then be used as a key in a binary tree i.e., building a routing table.

The following is a working instance. It works on network order (i.e., first bit is on the left) so if issue #29 is fixed, this will need to be modified.

```
module Net.IPv4.RangeBits where
import Net.IPv4
import Net.IPv4.Range
import Data.Bits
import Data.Maybe (fromMaybe)

-- Run a bitwise operation on the IPv4 component of a range
-- For ranges with different sizes we "promote operands" to the larger network of the two.
-- This is the same as dealing with integers i.e., 1 & 1001 is actually 0001 & 1001 = 0001
-- Note: any host bits set are added to the network bits if lengths differ
rangeBitwise :: (IPv4 -> IPv4 -> IPv4) -> IPv4Range -> IPv4Range -> IPv4Range
rangeBitwise fun l r = IPv4Range ip len
  where ip = (ipv4RangeBase l) `fun` (ipv4RangeBase r)
        len = maximum [ipv4RangeLength l, ipv4RangeLength r]

-- Run a bitwise operation on the IP (normalising it after), keeping the same length
rangeRebase :: (IPv4 -> IPv4) -> IPv4Range -> IPv4Range
rangeRebase fun r = IPv4Range (fun $ ipv4RangeBase r) (ipv4RangeLength r)

-- Bit operations for an IPv4Range operate on the whole IP passed (not just the network portion specified) but do not
-- modify the length / mask. We do this because an IP range may be overloaded to specify a network and a host (decided
-- at creation)
instance Bits IPv4Range where
  (.&.) = rangeBitwise (.&.)
  (.|.) = rangeBitwise (.|.)
  xor = rangeBitwise xor
  complement = rangeRebase complement
  shift r i = rangeRebase (flip shift i) r
  rotate r i = rangeRebase (flip rotate i) r
  bitSize ip = fromMaybe 0 (bitSizeMaybe ip)
  bitSizeMaybe = Just . fromIntegral . ipv4RangeLength
  isSigned = isSigned . ipv4RangeBase
  testBit ip i = testBit (ipv4RangeBase ip) $ 31 - i
  bit i = IPv4Range (bit $ 31 - i) (fromIntegral $ i + 1)
  popCount = popCount . ipv4RangeBase

instance FiniteBits IPv4Range where
  finiteBitSize = fromIntegral . ipv4RangeLength
```

**Comment by LemonIceScream** [comment]
Below is also RangeBitsSpec.hs for the test cases:

```
module Net.IPv4.RangeBitsSpec (spec) where
import Prelude hiding (any)
import Test.Hspec
import Data.Bits
import Data.Maybe (fromJust)
import Net.IPv4
import Net.IPv4.Range
import Net.IPv4.RangeBits

default_ :: IPv4Range
default_ = IPv4Range any 0

spec :: Spec
spec = parallel $ do
    describe "IPv4Range" $ do
        context "Bits" $ do
            it "bitSizeMaybe" $ do
                bitSizeMaybe (IPv4Range any 0) `shouldBe` Just 0
                bitSizeMaybe (IPv4Range loopback 15) `shouldBe` Just 15
                bitSizeMaybe (IPv4Range broadcast 32) `shouldBe` Just 32
            it "bit" $ do
                bit 0 `shouldBe` IPv4Range (ipv4 128 0 0 0) 1
                bit 1 `shouldBe` IPv4Range (ipv4 64 0 0 0) 2
                bit 31 `shouldBe` IPv4Range (ipv4 0 0 0 1) 32
                bit (fromJust $ bitSizeMaybe $ IPv4Range any 0) `shouldBe` IPv4Range (ipv4 128 0 0 0) 1
            it "testBit" $ do
                fmap (testBit default_) [0..31] `shouldBe` replicate 32 False
                testBit default_ 16 `shouldBe` False
                testBit (IPv4Range broadcast 16) 24 `shouldBe` True
        context "FiniteBits" $ do
            it "finiteBitSize" $ do
                finiteBitSize (IPv4Range any 16) `shouldBe` 16
                finiteBitSize default_ `shouldBe` 0
            it "bool iteration" $ do
                let net = range (ipv4 223 255 254 0) 24
                    b = [ True, True, False, True, True, True, True, True
                        , True, True, True, True, True, True, True, True
                        , True, True, True, True, True, True, True, False]
                fmap (testBit net) [0..((fromJust $ bitSizeMaybe net) - 1)] `shouldBe` b
```

**Comment by andrewthad** [comment]
This is super cool. I never considered that instance. I'll take a PR adding this.

**Comment by chessai** [comment]
Can this be closed because of PR #32?

**Comment by andrewthad** [comment]
Yes.
